Comrv baremetal

ronen-haen-sndk opened the pull request:
add a dedicated stack to comrv calls

ronen-haen-sndk committed:
fix build for rtos demo

ronen-haen-sndk committed:
merge with master

ronen-haen-sndk committed:
add comrv entry implementation

ronen-haen-sndk committed:
use a register save caller ra offset and token

ronen-haen-sndk committed:
remove License directory

ronen-haen-sndk committed:
fix build warning

ronen-haen-sndk committed:
make comrv compile with llvm

ronen-haen-sndk committed:
add COM-RV stack support

ronen-haen-sndk committed:
add comrv stack support

ronen-haen-sndk committed:
fix build

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+    /* we need to load the caller token to detrmine if the RA is an offset
+       or an actual return address; first we need to unlink the top of the
+       com-rv stack frame, return it back to the stack frame pool and get
+       read the caller token (in previous frame it will bethe  callee token) */
```
small typo

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+    jalr    t5
+    /* load the return address */
+    m_LOAD  ra, D_COMRV_CALLER_RA_STACK_OFFSET(t3)
+    /* we need to load the caller token to detrmine if the RA is an offset
```
offset should be; res address or ovl address

ronen-haen-sndk replied:
I'm using 'offset' in all fields and it refers to where in the stack it is located

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+       com-rv stack frame, return it back to the stack frame pool and get
+       read the caller token (in previous frame it will bethe  callee token) */
+    /* load the offset to previous stack frame */
+    lh      t5, D_COMRV_PREV_STACK_FRAME(t3)
```
add macro fir LH, due to the fact the 64 lh will load 4 bytes

ronen-haen-sndk replied:
no need for a macro - lh will always load half word (2 bytes)

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+* @author Ronen Haen
+* @date   21.06.2019
+* @brief  The file implements the COM-RV entry function
+*
```
add table explain t3, t4, t5

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+       t2 will hold the offset to the previous free stack frame */
+    sub     t4, t4, t3
+    /* save the offset to the previous free stack frame */
+    sh      t4, D_COMRV_PREV_STACK_FRAME(t3)
```
add macro for SH

ronen-haen-sndk replied:
no need for a macro as sh always stores 2 bytes (in rv32 and rv64)

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+       the previous stack frame) */
+    m_LOAD  t5, D_COMRV_CALLEE_TOKEN_STACK_OFFSET(t3)
+    /* t2 will hold the token type - address/token */
+    andi    t2, t5, 1
```
add macro : check token

ronen-haen-sndk replied:
added D_COMRV_TOKEN_INDICATION

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+       token search and load */
+    beqz    t2, comrv_exit_ret_to_caller
+    /* change sp for temporary storage */
+    m_ADDI  sp, sp, -D_COMRV_STACK_SIZE
```
change to frame_size

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,180 @@
+       we just provide it a temp location on the stack */
+    m_ADDI  a0, sp, D_COMRV_ARG_STACK_OFFSET_TMP
+    /* search if the token (t5) is already loaded */
+    jal     comrvGerFunctionAddressFromToken
```
on comment: fix ger to get
on code: think on return the "real address" and not doing manipulation for offset. do it inside the function

oferShinaar reviewed (commented):
added comments to review

ronen-haen-sndk committed:
backup

ronen-haen-sndk committed:
backup - first real comrv call works

ronen-haen-sndk committed:
move offset calculation from .S to .c file + add main implementation

ronen-haen-sndk committed:
add comments

ronen-haen-sndk committed:
linker script- change .ovlgrp.x to .ovlgrpdata
comrv.c - modify comments, rename some macros and handle return offset with multi group
comrv_entry.S - add overlay support with function pointer (thunks)
SConstruct - move here the -ffixed-regX option

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,188 @@
+    /* temporary save a0 in the next free stack frame */
+    m_STORE a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* current comrv stack frame is needed when calling a new overlay function so */
+    mv      a0, t3
```
- no need to move t3 to a0 since t3 is reserved (like global)

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,188 @@
+    /* Add the indication we are calling a new overlay function
+       this indication is required to calculate the correct offset
+       when returning to an overlay function */
+    ori     a0, a0, D_COMRV_PROFILING_INVOKE_VAL
```
- define new label without profiling...
- name: 
#define COMRV_INVOKE_CALLE_BIT_0  (1)
#define COMRV_RET_CALLER_BIT_0  (0)

ronen-haen-sndk replied:
Done

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,188 @@
+    m_LOAD  a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* adjust back sp - memory not needed anymore */
+    m_ADDI  sp, sp, D_COMRV_STACK_FRAME_SIZE
+comrv_invoke_callee:
```
do spacing between labels

ronen-haen-sndk replied:
done

oferShinaar commented on WD-Firmware/comrv/comrv_entry.S:
```diff
@@ -0,0 +1,188 @@
+comrv_invoke_callee:
+    /* call the function */
+    jalr    t5
+comrv_skip_thunk:
```
change to igonr_callee_thunk_stack_frame

ronen-haen-sndk replied:
I used comrv_igonr_caller_thunk_stack_frame as when returning, the thunk is the caller and not the callee

oferShinaar commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,118 @@
+/* 
+* SPDX-License-Identifier: Apache-2.0
```
[OS]: we stopped on viewing comrv_entry.s

ronen-haen-sndk committed:
code review changes
- make comrvGetAddressFromToken to receive no args
- use D_COMRV_INVOKE_CALLEE_BIT_0 instead of D_COMRV_PROFILING_INVOKE_VAL and D_COMRV_RET_CALLER_BIT_0
- +additional minor changes

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,629 @@
 * include files
 */
+#include "common_types.h"
+#include "string.h"
```
remove string.h

rhwd replied:
it is there due to memset usage

ronen-haen-sndk replied:
done in e1b829c

ronen-haen-sndk committed:
add missing thunk field to comrvTokenFields_t

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+/* overlay token */
+typedef union comrvOverlayToken
+{
+  u32_t              value;
```
add TODO comment: check without bitfield or t5 direct access

ronen-haen-sndk replied:
done in 4a8e389

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+} comrvOverlayToken_t;
+
+/* overlay token */
+typedef struct comrvPropertiesFields
```
same comment as line 130

ronen-haen-sndk replied:
check types copy from bit fields and see if code is smaller

ronen-haen-sndk replied:
done in 4a8e389

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+/* overlay token */
+typedef struct comrvPropertiesFields
+{
+  u08_t locked:1;
```
check structure fields coding conventions

ronen-haen-sndk replied:
done in 7829083

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+  comrvTokenFields_t fields;
+} comrvOverlayToken_t;
+
+/* overlay token */
```
change the comment

ronen-haen-sndk replied:
done in 7829083

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+{
+  u08_t locked:1;
+  u08_t data:1;
+  u08_t size:3;
```
check if need to be 4
change the size to a better naming

ronen-haen-sndk replied:
done in 7829083

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+} comrvEntryProperties_t;
+
+/* overlay token entry */
+typedef struct comrvOverlayHeapEntry
```
change "Heap" to "Cache" all over the place

ronen-haen-sndk replied:
done in 7829083

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+#endif /* D_COMRV_EVICTION_LRU */
+  comrvEntryProperties_t properties;
+  comrvOverlayToken_t    token;
+  void*                  pAddress;
```
change the name to: fixed entry address

ronen-haen-sndk replied:
done in 7829083

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+typedef struct comrvOverlayHeapEntry
+{
+#ifdef D_COMRV_EVICTION_LRU
+  u08_t prevIndex;
```
add comments describing the fields
add a profile compile switch

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+} comrvHeapEntry_t;
+
+/* com-rv control block */
+typedef struct comrvCB
```
we stopped the meeting here!

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
 * external prototypes
 */
+extern void  comrv_entry               (void);
+extern void  comrvMemcpyHook           (void* pDest, void* pSrc, u32_t sizeInBytes);
```
add comment: user hook

ronen-haen-sndk replied:
done in 7829083

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
 /**
 * external prototypes
 */
+extern void  comrv_entry               (void);
```
comrv_entry: try to remove it or add a comment that this is not user "need to write" hook.
[try global public on the asm]

ronen-haen-sndk replied:
'public'not supported by gas
added comment in 4b2b10c

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
 /**
 * global variables
 */
+comrvStackFrame_t* pStackStartAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_START__;
```
need to be giving us an input to init

ronen-haen-sndk replied:
done in 0c0903b

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
 */
+comrvStackFrame_t* pStackStartAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_START__;
+comrvCB_t*         pComrvCB;
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
```
add comments

ronen-haen-sndk replied:
done in 0c0903b

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+comrvStackFrame_t* pStackStartAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_START__;
+comrvCB_t*         pComrvCB;
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
+comrvOverlayToken_t *pOverlayMultiGroupTokensTable = (comrvOverlayToken_t*)&overlayMultiGroupTokensTable;
```
will be deprecated

ronen-haen-sndk replied:
will be implemented once we get the toolchain support (placing the table in overlay group #0)

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+void comrvInit(void)
+{
+   u08_t              i;
+   void*              pBaseAddress = &__OVERLAY_SEC_START__;
```
- create 2 num-symbols on linker and use them on the new global which will be used for engine data

-  get comrv_sp from the input struct that the user provide. size will be in #define in the Comrv.config.h for numbers of entries. user can change it if he wants. the stack itself will be in global (dccm)

ronen-haen-sndk replied:
The first bullet can't be implemented -
"error: variable length array declaration not allowed at file scope"
This is due to the fact that in compile time the actual size isn't known yet
Second bullet done in 0c0903b

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+   pComrvCB->numOfOverlayEntries = D_COMRV_NUM_OF_OVERLAY_ENTRIES;
+   /* determine the location of comrv overlay heap structure */
+   //pComrvCB->overlayHeap = (comrvHeapEntry_t*)((u08_t*)pComrvCB - (pComrvCB->numOfOverlayEntries*sizeof(comrvHeapEntry_t)));
+#ifdef D_COMRV_EVICTION_LRU
```
we stop the meeting here!

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+#endif /* D_COMRV_EVICTION_LRU */
+   pStackStartAddr = (comrvStackFrame_t*)&pComrvCB->overlayHeap[pComrvCB->numOfOverlayEntries];
+   /* Init comrv stack - mark the last stack frame */
+   pStackStartAddr->offsetPrevFrame = D_COMRV_END_OF_STACK;
```
add todo comment inside the code: make stack check for the comrv stack
add an api "comrv get status" - run till edge of stack

ronen-haen-sndk replied:
done in 0f3602c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+   /* initialize all stack entries */
+   for (pStackStartAddr++ ; pStackStartAddr < pStackEndAddr ; pStackStartAddr++)
+   {
+      pStackStartAddr->offsetPrevFrame = (s16_t)-sizeof(comrvStackFrame_t);
```
check this without casting

ronen-haen-sndk replied:
if we remove the casting we get the following:
"warning: implicit conversion from 'unsigned int' to 's16_t' (aka 'short') changes value from 4294967284 to -12 [-Wconstant-conversion]"

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+   for (pStackStartAddr++ ; pStackStartAddr < pStackEndAddr ; pStackStartAddr++)
+   {
+      pStackStartAddr->offsetPrevFrame = (s16_t)-sizeof(comrvStackFrame_t);
+      pStackStartAddr->calleeToken = 0;
```
add this only to tail

ronen-haen-sndk replied:
done in 0f3602c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+   /* set the address of COMRV stack and initialize it */
+   pStackStartAddr--;
+   M_COMRV_SET_STACK_ADDR(pStackStartAddr);
+   pStackStartAddr->offsetPrevFrame = D_COMRV_END_OF_STACK;
```
check duplicate with 241

ronen-haen-sndk replied:
done in ce455df

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+   }
+
+#ifndef D_COMRV_USE_OS
+   /* set the address of COMRV stack and initialize it */
```
replace with function

ronen-haen-sndk replied:
done in b080aea

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,645 @@
+      pStackStartAddr->calleeToken = 0;
+   }
+
+#ifndef D_COMRV_USE_OS
```
we stopped here!
will talk about comrv init after OS started

rhwd replied:
N/A

ronen-haen-sndk committed:
changes following CR

ronen-haen-sndk committed:
add comment to hook functions

ronen-haen-sndk committed:
remove NULL #define

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+}
+
+/**
+* Search if current overlay token is already loaded
```
update comments

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+#ifdef D_COMRV_CRC
+   u32_t               uiCrc;
+#endif /* D_COMRV_CRC */
+#ifdef D_COMRV_FW_PROFILING
```
change profiling to instrumentation

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+   u16_t               usSelectedMultiGroupEntry;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* read the requested token value */
```
add (t5) to the comment

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+   if (ucIsInvoke == D_COMRV_INVOKE_CALLEE_BIT_0)
+   {
+      /* clear the invoke indication from pComrvStackFrame address */
+      pComrvStackFrame = (comrvStackFrame_t*)((u32_t)pComrvStackFrame & (~D_COMRV_PROFILING_INVOKE_VAL));
```
check if we can remove the casting

ronen-haen-sndk replied:
can't be removed; get the following compilation error -
"invalid operands to binary expression"

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+   /* search for a multi-group overlay token */
+   else
+   {
+      pAddress = NULL;
```
remove this line

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+         /* search for the token */
+         pAddress = comrvSearchForLoadedOverlayGroup(pOverlayMultiGroupTokensTable[ucEntryIndex]);
+      /* continue the search as long as the group wasn't found and we have additional tokens */
+      } while ((pAddress == NULL) && (pOverlayMultiGroupTokensTable[ucEntryIndex].ucValue != 0));
```
add ++ to ucEntryIndex

ronen-haen-sndk replied:
done in 7b38d21

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+   }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* get the group size */
```
we stopped here

ronen-haen-sndk commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,662 @@
+         /* search for the token */
+         pAddress = comrvSearchForLoadedOverlayGroup(pOverlayMultiGroupTokensTable[ucEntryIndex]);
+      /* continue the search as long as the group wasn't found and we have additional tokens */
+      } while ((pAddress == NULL) && (pOverlayMultiGroupTokensTable[ucEntryIndex].ucValue != 0));
```
fix type of ucValue  'uc'

ronen-haen-sndk replied:
done in 7b38d21

ronen-haen-sndk committed:
changes following CR

ronen-haen-sndk committed:
CR changes:
1. add comrvGetStatus function
2. in comrvInit, move pStackStartAddr->calleeToken = 0; to be done only for the last item of the stack

ronen-haen-sndk committed:
CR changes:
1. in comrvInit remove duplication of pStackStartAddr->offsetPrevFrame = D_COMRV_END_OF_STACK (initially it was done in the wrong frame)

ronen-haen-sndk committed:
CR changes:
1. add comrvInitApplicationStack function to initialize comrv application stack

ronen-haen-sndk committed:
remove volatile'

ronen-haen-sndk committed:
CR changes:
1. remove comrv stack from ld script
2. add argument to comrvInit
3. make the comrv stack and comrv CB globals instead of receiving them from ld
4. add build error in case no eviction algorithm selected
5. add comments + vars naming fixes

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+      /* disable ints */
+      // TODO: disable ints
```
add COMRV macro for enter critical

ronen-haen-sndk replied:
done in 440b0c4

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+            /* get the candidate entry */
+            pEntry = &stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]];
+            /* calc the source address */
+            pAddress = ((u08_t*)pEntry->pFixedEntryAddress + M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(pEntry->unToken));
```
add more comments details.

ronen-haen-sndk replied:
done in 90f5569

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+            pAddress = ((u08_t*)pEntry->pFixedEntryAddress + M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(pEntry->unToken));
+            /* perform code copy */
+            comrvMemcpyHook(pEntry->pFixedEntryAddress, pAddress,
+                  stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress - pAddress);
```
1. stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress 
  assign local "meaningful" arg.
2. ucEvictCandidateList[ucEntryIndex] should be ucEvictCandidateList[ucEntryIndex+1]
3. write : todo comment for consider use calculations instead of : ((u08_t*)pEntry->pFixedEntryAddress  since it is fixed and load are more expansive then cal.

ronen-haen-sndk replied:
done in 90f5569

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+                  stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress - pAddress);
+
+            /* after code copy we need to align the entries structures */
+            for(ucIndex = ucEvictCandidateList[ucEntryIndex] ; ucIndex < ucEvictCandidateList[ucEntryIndex+1] ; ucIndex++)
```
do: for () with space

ronen-haen-sndk replied:
done in 90f5569

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+            /* after code copy we need to align the entries structures */
+            for(ucIndex = ucEvictCandidateList[ucEntryIndex] ; ucIndex < ucEvictCandidateList[ucEntryIndex+1] ; ucIndex++)
+            {
+               pEntry = &stComrvCB.stOverlayCache[ucIndex + stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits];
```
1. add more comments 
2. enb int: todo macro for exit critical

ronen-haen-sndk replied:
done in 90f5569

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+               stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = pEntry->unToken.uiValue;
+#ifdef D_COMRV_EVICTION_LRU
+               stComrvCB.stOverlayCache[pEntry->ucPrevIndex].ucNextIndex = ucIndex;
+               stComrvCB.stOverlayCache[pEntry->ucNextIndex].ucPrevIndex = ucIndex;
```
1. could be error on next and prev are the same
2. make union

ronen-haen-sndk replied:
done in 8701f9d

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+                  comrvNotificationHook(D_COMRV_OVL_DATA_DEFRAG_ERR, unToken.ucValue);
+               }
+#endif /* D_COMRV_OVL_DATA_SUPPORT */
+               stComrvCB.stOverlayCache[ucIndex].unProperties = pEntry->unProperties;
```
make union

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+            }
```
end comment for end of loop

ronen-haen-sndk replied:
done in 90f5569

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+               pEntry = &stComrvCB.stOverlayCache[ucIndex + stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits];
+#ifdef D_COMRV_OVL_DATA_SUPPORT
+               /* an overlay data is present when handling defragmentation */
+               if (pEntry->unProperties.ucData)
```
do in assert macro

ronen-haen-sndk replied:
can't have here assert since in the next line we have M_COMRV_EXIT_CRITICAL_SECTION();

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+      /* if the requested token is a multi-group token */
+      if (unToken.stFields.multiGroup != 0)
+      {
+         /* for now we take the first token in the list of tokens */
```
add todo comment for smart way to get multi group loading

ronen-haen-sndk replied:
done in ea2e603

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+      {
+         /* for now we take the first token in the list of tokens */
+         unToken = pOverlayMultiGroupTokensTable[unToken.stFields.overlayGroupID];
+         /* save the selected multi group entry */
```
explain why we save the entry in the comment

ronen-haen-sndk replied:
done in 14d152f

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,767 @@
+            }
+            ucEntryIndex++;
+         }
+         ucNumOfEvictionCandidates++;
```
we stop the meeting here!

ronen-haen-sndk committed:
CR changes:
1. add empty macros for enter/exit critical section

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      else if (ucNumOfEvictionCandidates == 0)
+      {
+         M_COMRV_EXIT_CRITICAL_SECTION();
+         comrvNotificationHook(D_COMRV_NO_AVAILABLE_ENTRY_ERR, unToken.uiValue);
```
implementation of error_hook
implementation of instrumentation_hook

ronen-haen-sndk replied:
done in d17a18a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+         ucNumOfEvictionCandidates++;
+      }
+      /* this means the end user locked all entries */
+      else if (ucNumOfEvictionCandidates == 0)
```
make this if first

ronen-haen-sndk replied:
did reduce code size in 2 bytes

ronen-haen-sndk replied:
done in bc69d9f

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      }
+      ucIndex = ucEvictCandidateList[ucEntryIndex];
+      /* update the entry access */
+      comrvUpdateHeapEntryAccess(ucIndex);
```
change to UpdateCacheEvectionParams

ronen-haen-sndk replied:
done in c55598a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = unToken.uiValue;
+      /* update the heap entry properties with the group size */
+      stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits = usOverlayGroupSize;
+      /* if evicted size is larger than requested size we need to update the remaining tail*/
```
rephrase this comment (update the CB remaining space)

ronen-haen-sndk replied:
done in c55598a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      if (ucSizeOfEvictionCandidates != 0)
+      {
+         ucEntryIndex += usOverlayGroupSize;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits = ucSizeOfEvictionCandidates;
```
add comment

ronen-haen-sndk replied:
done in c55598a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+#ifdef D_COMRV_EVICTION_LRU
+         stComrvCB.stOverlayCache[stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex].ucNextIndex = stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex           = ucIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex           = D_COMRV_LRU_FIRST_ITEM;
```
check if this should MRU

ronen-haen-sndk replied:
no,  now it is the first lru as it is now considered 'evicted/empty'

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+         stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex           = D_COMRV_LRU_FIRST_ITEM;
+         /* mark the group ID so that it won't pop in the next search */
+         stComrvCB.stOverlayCache[ucEntryIndex].unToken.uiValue       = D_COMRV_ENTRY_TOKEN_INIT_VALUE;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.ucValue &= D_COMRV_ENTRY_PROPERTIES_RESET_MASK;
```
double check if the size value is correct

ronen-haen-sndk replied:
done in c55598a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      {
+         ucEntryIndex += usOverlayGroupSize;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits = ucSizeOfEvictionCandidates;
+#ifdef D_COMRV_EVICTION_LRU
```
move this after load

rhwd replied:
I think we concluded that moving this isn't correct

ronen-haen-sndk replied:
we decided that it is not useful to move the code - we'll handle context switch

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+         stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex           = D_COMRV_LRU_FIRST_ITEM;
+         /* mark the group ID so that it won't pop in the next search */
+         stComrvCB.stOverlayCache[ucEntryIndex].unToken.uiValue       = D_COMRV_ENTRY_TOKEN_INIT_VALUE;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.ucValue &= D_COMRV_ENTRY_PROPERTIES_RESET_MASK;
```
take these lines out of the "if"

ronen-haen-sndk replied:
done in c55598a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+#ifdef D_COMRV_EVICTION_LRU
+         stComrvCB.stOverlayCache[stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex].ucNextIndex = stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex           = ucIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex           = D_COMRV_LRU_FIRST_ITEM;
```
Add LRU to unNextIndex, ucPrevIndex

ronen-haen-sndk replied:
done in 40ea73e

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+         /* mark the group ID so that it won't pop in the next search */
+         stComrvCB.stOverlayCache[ucEntryIndex].unToken.uiValue       = D_COMRV_ENTRY_TOKEN_INIT_VALUE;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.ucValue &= D_COMRV_ENTRY_PROPERTIES_RESET_MASK;
+         stComrvCB.ucLruIndex = ucIndex;
```
ucLruIndex add "top"

ronen-haen-sndk replied:
We agreed that this isn't needed as the name specifies the meaning - lru and mru

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      /* it is safe now to get new requests */
+      M_COMRV_EXIT_CRITICAL_SECTION();
+      /* the group size in bytes */
+      usOverlayGroupSize <<= 9;
```
you know what to do

ronen-haen-sndk replied:
done in ea2e603

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      /* the group size in bytes */
+      usOverlayGroupSize <<= 9;
+      /* now we can load the overlay group */
+      pAddress = comrvLoadOvlayGroupHook(M_COMRV_GET_GROUP_OFFSET_IN_BYTES(stComrvCB.stOverlayCache[ucIndex].unToken),
```
replace functions arguments with structure

ronen-haen-sndk replied:
done in ae5bf40

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      /* if group wasn't loaded */
+      if (pAddress == NULL)
+      {
+         comrvNotificationHook(D_COMRV_LOAD_ERR, unToken.uiValue);
```
pass structure to error and instrumentation hooks

ronen-haen-sndk replied:
done in d17a18a

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      pAddress = comrvLoadOvlayGroupHook(M_COMRV_GET_GROUP_OFFSET_IN_BYTES(stComrvCB.stOverlayCache[ucIndex].unToken),
+            stComrvCB.stOverlayCache[ucIndex].pFixedEntryAddress, usOverlayGroupSize);
+      /* if group wasn't loaded */
+      if (pAddress == NULL)
```
check if llvm has __rarely attribute

ronen-haen-sndk replied:
I didn't find - pending reply from embecosm

ronen-haen-sndk replied:
done in dfe73e0

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+#ifdef D_COMRV_CRC
+      /* calculate crc */
+      uiCrc = comrvCrcCalcHook(pAddress, usOverlayGroupSize-sizeof(u32_t));
+      if (uiCrc != *((u08_t*)pAddress + (usOverlayGroupSize-sizeof(u32_t))))
```
add macro and pass address, size, target result

ronen-haen-sndk replied:
done in c68affa

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      uiCrc = comrvCrcCalcHook(pAddress, usOverlayGroupSize-sizeof(u32_t));
+      if (uiCrc != *((u08_t*)pAddress + (usOverlayGroupSize-sizeof(u32_t))))
+      {
+         comrvNotificationHook(D_COMRV_CRC_CHECK_ERR, unToken.ucValue);
```
look at get status function

ronen-haen-sndk replied:
done in 3b61f8c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+
+   /* group is now loaded to memory so we can return the address of the data/function */
+   return (void*)((u08_t*)pAddress + usOffset);
```
check if you need (void*) at this point

ronen-haen-sndk replied:
done in 9676d72

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+      /* In multi group token we need to get the offset from previously loaded token */
+      if (pComrvStackFrame->usCalleeMultiGroupTableEntry)
+      {
+         /* we now are at the point of loading a multi-group toke so we need to take the
```
token instead of toke

ronen-haen-sndk replied:
done in 64dd0f8

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+         /* get the token group size */
+         usOverlayGroupSize = M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(unToken);
+         /* calculate the actual return offset */
+         usOffset += ((u32_t)(pComrvStackFrame->uiCalleeToken) - uiTemp) & (--usOverlayGroupSize);
```
clear the value of this field usCalleeMultiGroupTableEntry at this point

ronen-haen-sndk replied:
done in 64dd0f8

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,765 @@
+* @param requestedEvictionSize - size requested for eviction; expressed in
+*                                units of D_COMRV_OVL_GROUP_SIZE_MIN
+*        pEvictCandidatesList  - output eviction candidate list of comrv heap indexes
+* @return number of eviction candidates in the output list
```
we stopped here

ronen-haen-sndk committed:
CR changes:
1. add comments
2. stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress
assign local meaningful arg
3. cEvictCandidateList[ucEntryIndex] should be ucEvictCandidateList[ucEntryIndex+1]

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+
+#ifdef D_COMRV_EVICTION_LRU
+   /* get the first lru entry */
+   ucEntryIndex = stComrvCB.ucLruIndex;
```
change the name to meet "global" code convention

ronen-haen-sndk replied:
done in 42eaa83

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+   u32_t uiEvictCandidateMap[D_COMRV_EVICT_CANDIDATE_MAP_SIZE], uiCandidates;
+
+   /* first lets clear the uiCandidates list */
+   memset(uiEvictCandidateMap, 0, sizeof(u32_t)*D_COMRV_EVICT_CANDIDATE_MAP_SIZE);
```
"D_COMRV_EVICT_CANDIDATE_MAP_SIZE" link this define to the configuration data
replace "memset" with "comrv memset"

ronen-haen-sndk replied:
done in e1b829c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+   do
+   {
+      /* verify the entry isn't locked */
+      if (stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucLocked != D_COMRV_ENTRY_LOCKED)
```
"stComrvCB.stOverlayCache[ucEntryIndex]" for this element use a local variable

ronen-haen-sndk replied:
done in 830d2fc

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+   /* loop as long as we didn't get to the requested eviction size or we reached end of the list
+      (means that all entries are locked) */
+   } while (ucAccumulatedSize < ucRequestedEvictionSize && ucEntryIndex != D_COMRV_LRU_LAST_ITEM);
+
```
add a condition check if accumulated size < requested size, add an error "not enough entries"

ronen-haen-sndk replied:
done in 5e17026

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+         /* get the lsb that is set */
+         pEvictCandidatesList[ucIndex] = uiCandidates & (-uiCandidates);
+         /* subtract the lsb that is set */
+         uiCandidates -= pEvictCandidatesList[ucIndex]--;
```
break this into two lines and add comment why do you need "--"

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+         pEvictCandidatesList[ucIndex] = uiCandidates & (-uiCandidates);
+         /* subtract the lsb that is set */
+         uiCandidates -= pEvictCandidatesList[ucIndex]--;
+         pEvictCandidatesList[ucIndex++] += ucEntryIndex*D_COMRV_DWORD_IN_BITS;
```
same as previous comment

ronen-haen-sndk replied:
done in d3da7a7

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+         /* subtract the lsb that is set */
+         uiCandidates -= pEvictCandidatesList[ucIndex]--;
+         pEvictCandidatesList[ucIndex++] += ucEntryIndex*D_COMRV_DWORD_IN_BITS;
+      }
```
add an assert for loops greater than 8

ronen-haen-sndk replied:
done in 0f94b53

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+   return ucNumberOfCandidates;
+}
+
+/**
```
we stopped here

rhwd commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+   for (ucIndex = 0 ; ucIndex < D_COMRV_NUM_OF_CACHE_ENTRIES ; ucIndex++)
+   {
+      /* initially each entry points to the previous and next neighbor cells */
+      stComrvCB.stOverlayCache[ucIndex].ucPrevIndex          = ucIndex-1;
```
setting stComrvCB.stOverlayCache[ucIndex] to a local pointer increases the code size of comrvInit() by 2 bytes

ronen-haen-sndk commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,777 @@
+      pBaseAddress = ((u08_t*)pBaseAddress + D_COMRV_OVL_GROUP_SIZE_MIN);
+   }
+   /* mark the last entry in the LRU list */
+   stComrvCB.stOverlayCache[ucIndex-1].ucNextIndex = D_COMRV_LRU_LAST_ITEM;
```
changing this - 
stComrvCB.stOverlayCache[ucIndex-1].unLru.stFields.typNextIndex 
with this -
pEntry->unLru.stFields.typNextIndex
increases the code size in 4 bytes - so for this line I didn't use pEntry->unLru.stFields.typNextIndex

ronen-haen-sndk committed:
CR changes:
1. rename remaining 'heap' to 'cache'
2. modify lru fields to be union
3. in comrvInit decrease code size by using local variable pCacheEntry
4. in comrvUpdateCacheEntryAccess make code more readable by adding local pointer to point to the updated entry (&stComrvCB.stOverlayCache[ucEntryIndex])

ronen-haen-sndk committed:
add comment to usSelectedMultiGroupEntry

ronen-haen-sndk committed:
CR changes:
1. remove unused #define
2. reorder condition check for ucNumOfEvictionCandidates

ronen-haen-sndk committed:
CR changes:
1. rename comrvUpdateCacheEntryAccess to comrvUpdateCacheEvectionParams
2. fix properties size field in case the evicted size was bigger than the requested size
3. comments

ronen-haen-sndk committed:
CR changes:
remane typNextIndex to typNextLruIndex and typPrevIndex to typPrevLruIndex

ronen-haen-sndk committed:
CR changes:
1. comments
2. add macro to convert group size to bytes

ronen-haen-sndk committed:
CR changes:
1. change comrvLoadOvlayGroupHook() to recieve structure instead of several args

ronen-haen-sndk committed:
CR changes:
1. in comrvGetAddressFromToken remove (void*) casting in return value

ronen-haen-sndk committed:
CR changes:
1. clear the usCalleeMultiGroupTableEntry field so it won't be reused
2. fix typo

ronen-haen-sndk committed:
CR changes:
1. rename stComrvCB to g_stComrvCB
2. rename stComrvStackPool to g_stComrvStackPool

ronen-haen-sndk committed:
CR changes:
1. change D_COMRV_EVICT_CANDIDATE_MAP_SIZE to use the values from the comrv_config.h file
2. add comrv version memcpy instead of using the library function

ronen-haen-sndk committed:
CR changes:
in comrvGetEvictionCandidates add a local variable to hold stComrvCB.stOverlayCache[ucEntryIndex]

ronen-haen-sndk committed:
fix build issue

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,853 @@
+* @return if the token is loaded the return value is set to the loaded address
+*         otherwise NULL
+*/
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t unToken)
```
return the index instead of the address and remove the eviction params update

ronen-haen-sndk replied:
done in 2fe7133

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,853 @@
+      /* update ucEntryIndex previous index */
+      pCacheEntry->unLru.stFields.typPrevLruIndex = g_stComrvCB.ucMruIndex;
+      /* update ucEntryIndex next index - last item (MRU)*/
+      pCacheEntry->unLru.stFields.typNextLruIndex = D_COMRV_LRU_LAST_ITEM;
```
change to MRU

rhwd replied:
This is OK, I'm using the terms first LRU (LRU item) and last LRU (MRU item)

ronen-haen-sndk replied:
done in a30593c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,853 @@
+         /* update the global lru index */
+         g_stComrvCB.ucLruIndex = pCacheEntry->unLru.stFields.typNextLruIndex;
+         /* update the lru item with the previous item index */
+         g_stComrvCB.stOverlayCache[g_stComrvCB.ucLruIndex].unLru.stFields.typPrevLruIndex = D_COMRV_LRU_FIRST_ITEM;
```
check the define name LRU or MRU

ronen-haen-sndk replied:
done in a30593c

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,853 @@
+* @return none
+*/
+// TODO: ronen - use define for no inline
+__attribute__((noinline)) void comrvInitApplicationStack(void)
```
use your own macro for attributes
we stopped here

ronen-haen-sndk replied:
done in b24ddc4

ronen-haen-sndk committed:
CR changes:
1. remove comrvNotificationHook
2. add comrvInstrumentationHook
3. add comrvErrorHook
4. fix instrumentation #define names

ronen-haen-sndk committed:
CR changes:
use comrv macro for inline

rhwd commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,853 @@
+*/
+// TODO: ronen - use define for no inline
+__attribute__((noinline)) void comrvInitApplicationStack(void)
+{
```
we stopped here

ronen-haen-sndk committed:
CR changes:
1. change D_COMRV_LRU_LAST_ITEM to D_COMRV_MRU_ITEM
2. change D_COMRV_LRU_FIRST_ITEM to D_COMRV_LRU_ITEM

ronen-haen-sndk committed:
fix wrong arg passing to comrvErrorHook

ronen-haen-sndk committed:
in comrvSearchForLoadedOverlayGroup fix loop counting

ronen-haen-sndk committed:
CR changes:
1. fix overlay data condition check in comrvGetAddressFromToken
2. change comrvSearchForLoadedOverlayGroup to return an index instead of an address

ronen-haen-sndk committed:
CR changes:
in comrvGetEvictionCandidates, make the code more readble in the loop 'while (uiCandidates)'

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,884 @@
+   comrvStackFrame_t *pStackPool, *pStackFrame;
+
+   /* disable ints */
+   // TODO: disable ints
```
add "todo", with rtosal, implement condition check if scheduler is running or not, if running use sync api

ronen-haen-sndk replied:
done in ca8dbd0

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,884 @@
+}
+
+/**
+* comrv version of memcoy - copy dwords only
```
memset

ronen-haen-sndk replied:
done in ca8dbd0

nidalf commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,139 @@
+* include files
+*/
+#include "common_types.h"
+#include "string.h"
```
remove this include

ronen-haen-sndk replied:
done in ca8dbd0

nidalf commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,139 @@
+/**
+* globals
+*/
+extern void* __OVERLAY_GROUPS_START__;
```
change name to overlay storage start address

ronen-haen-sndk replied:
done in ca8dbd0

nidalf commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,139 @@
+*
+* @param
+*
+* @return none
```
change comment to @return destination

nidalf commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,139 @@
+*
+* @param
+*
+* @return none
```
comment typo, it is return destination not return none

ronen-haen-sndk replied:
done in ca8dbd0

nidalf commented on WD-Firmware/comrv/comrv_hooks.c:
```diff
@@ -0,0 +1,139 @@
+*
+* @return calculated CRC
+*/
+u32_t comrvCrcCalcHook(void* pAddress, u16_t memSizeInBytes)
```
wrap comrvCrcCalcHook function with compile switch like function comrvInstrumentationHook

ronen-haen-sndk replied:
the intention is to wrap the call to this function - there is already a CR comment for this in comrv.c

ronen-haen-sndk committed:
CR changes:
1. rename __OVERLAY_GROUPS_START__ to __OVERLAY_STORAGE_START__ADDRESS__
2. remove unnecessary #include
3. fix comments and typos

ronen-haen-sndk committed:
CR changes:
in comrvGetEvictionCandidates add a check if the  accumulated size < requestedsize and call comrvErrorHook accordingly

ronen-haen-sndk committed:
fix 'Load address misaligned' exception

ronen-haen-sndk committed:
CR changes:
invoke CRC via macro

ronen-haen-sndk committed:
fix a potential bug (due to latest changes) when returning from comrvGetEvictionCandidates we need to make sure we exit the critical section so we can't call directly to comrvErrorHook from comrvGetEvictionCandidates

ronen-haen-sndk committed:
CR changes:
in comrvGetEvictionCandidates assert if we reached a case where we have too many entries available

ronen-haen-sndk committed:
fix comment

ronen-haen-sndk committed:
CR changes:
add to comrvErrorHook call to comrvGetStatus

nidalf commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -18,18 +18,89 @@
 * @author Ronen Haen
 * @date   11.06.2019
-* @brief  The file defines the COM RV interfaces
+* @brief  The file defines the COM-RV interfaces
```
add check if llvm or not

ronen-haen-sndk replied:
done in 313ebf9

nidalf commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -18,18 +18,89 @@
+#endif /* D_COMRV_EVICTION_LRU */
+
+/* if no profile was set */
+#if D_COMRV_PROFILE==0
```
double check if define 0 equals is defined
End user should not choose the profile, pick it from the defines

ronen-haen-sndk replied:
done in 29e16d5

nidalf commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -18,18 +18,89 @@
+#define _OVERLAY_          __attribute__((overlaycall)) D_COMRV_NO_INLINE
+
+/* make sure eviction algorithm defined */
+#ifndef D_COMRV_EVICTION_LRU
```
make as in line 50

ronen-haen-sndk replied:
done in 29e16d5

nidalf commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,34 +18,753 @@
+
+#ifdef D_COMRV_VERIFY_INIT_ARGS
+   /* verify input parameters */
+   if ((!pInitArgs) || (pInitArgs->uiCacheSizeInBytes % D_COMRV_OVL_GROUP_SIZE_MIN))
```
add check if input size is aligned to defined size

ronen-haen-sndk replied:
done in 1e8cc05

nidalf commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -18,18 +18,89 @@
+#define D_COMRV_CRC_CHECK_ERR             4
+
+/* comrv defines */
+#define D_COMRV_PROFILE_BASE_IND          0xFC
```
change the name to instrumentation

ronen-haen-sndk replied:
done in 4ef975f

nidalf commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -39,6 +110,167 @@
+     actual loaded token entry in the multi group table */
+  u16_t usCalleeMultiGroupTableEntry;
+#endif /* d_comrv_multi_group_support */
+} comrvStackFrame_t;
```
+ #else

ronen-haen-sndk replied:
done in e546550

ronen-haen-sndk committed:
CR changes:
echo error if llvm isn't used to compile comrv module

ronen-haen-sndk commented on WD-Firmware/comrv/api_inc/comrv_api.h:
```diff
@@ -39,6 +114,167 @@
+}comrvErrorArgs_t;
+
+/* comrv instrumentation arguments */
+typedef struct comrvInstrumentationArgs
```
move all types to .h file

ronen-haen-sndk replied:
done in e546550

nidalf commented on WD-Firmware/comrv/api_inc/comrv_defines.h:
```diff
@@ -15,10 +15,10 @@
 * limitations under the License.
```
move here the defines from comrv_api.h

ronen-haen-sndk replied:
done in e546550

nidalf commented on WD-Firmware/comrv/loc_inc/comrv_asm_macros.h:
```diff
@@ -1,26 +1,40 @@
-/*
```
rename comrv_asm_macros.h to comrv_macros.inc

ronen-haen-sndk replied:
done in 5b8a6ec

ronen-haen-sndk committed:
CR changes:
1. fix D_COMRV_NUM_OF_CACHE_ENTRIES definition - not by the end user
2. add D_COMRV_EVICTION_POLICY definition - defined by the end user
3. typo

ronen-haen-sndk committed:
CR changes:rename PROFILE -> INSTRUMENTATION

ronen-haen-sndk committed:
CR changes:
1. move all types to comrv_types.h
2. move all #defines to comrv_defines.h
3. in comrvStackFrame_t add one padding of u16_t

ronen-haen-sndk committed:
CR changes:
rename comrv_asm_macro.h -> comrv_macros.inc

ronen-haen-sndk committed:
CR changes:
1. overlay cache - use linker symbols instead of global variables
2. overlay tables - use linker symbols instead of global variables
3. fix build issue due to rename to comrv_macros.inc

ronen-haen-sndk committed:
CR changes:
1. delete pFixedEntryAddress field from comrvCacheEntry_t; instead use linker symbols to calculate the address of the accessed cache entry
2. in main, add code to assist in future debug - set the mtvec register so we'll be able to break in case of exception

ronen-haen-sndk committed:
CR changes:
use __builtin_expect in errors to avoid branching

ronen-haen-sndk committed:
remove comment

ronen-haen-sndk committed:
comments

ronen-haen-sndk committed:
Fix potential bug due to wrong change I made in fragmentation:
1. when handling fragmentation I used the wrong index
2. make M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES use M_COMRV_GET_OVL_GROUP_SIZE
3. comments

ronen-haen-sndk committed:
remove unused local variable in comrvInit

ronen-haen-sndk committed:
CR changes:
add comment

ronen-haen-sndk committed:
CR changes:
add missing prefix
